### ISSUE TITLE
[Bug]: Vanilla top-p sampling can trigger CUDA device-side assert when cumulative probability stays below top_p

### ISSUE THREAD

**Issue by DhineshPonnarasan** [NVIDIA/TensorRT-LLM#19485]
### System Info

TensorRT-LLM commit: 7edb2b2a18
GPU: NVIDIA Tesla T4 (SM75), 15360 MiB
Driver: 580.178.04
CUDA reported by nvidia-smi: 13.0
Container: nvcr.io/nvidia/pytorch:26.02-py3
PyTorch: 2.11.0a0+eb65b36914.nv26.02
Python: 3.12.3
TensorRT: 10.15.1.29
Triton: 3.6.0
Platform: Linux

### Information

* [ ] The official example scripts
* [x] My own modified scripts

### Tasks

* [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
* [ ] My own task or dataset (give details below)

### Summary

A valid `top_p` value very close to 1.0 can trigger an out-of-bounds scatter in the vanilla top-p sampler.

The issue occurs when the FP32 cumulative probability sum remains slightly below `top_p` for an entire row due to finite-precision accumulation. In that case, `torch.searchsorted()` returns `vocab_size`, which is then used directly as a scatter index.

### Minimal reproduction

```python
import torch

from tensorrt_llm._torch.pyexecutor.sampler.ops.vanilla import (
    top_k_top_p_sampling_batch,
)

torch.manual_seed(0)

logits = torch.randn(2, 32000, device="cuda")

tokens, probs = top_k_top_p_sampling_batch(
    logits,
    temperature=1.0,
    top_p=0.9999999,
)

print(tokens)
```

Environment:

```bash
export TLLM_DEBUG_MODE=1
```

The above reproduces the failure on an NVIDIA Tesla T4.

### Reproduction rate

The failure reproduced 3/3 times in isolated processes with the same seed and inputs.

The same logic can also be isolated on CPU, where it produces an explicit out-of-bounds indexing error instead of a CUDA device-side assert.

### Expected behavior

`top_p=0.9999999` is a valid sampling parameter and should not cause an indexing error or CUDA device-side assert.

If the cumulative probability does not reach `top_p` because of finite-precision accumulation, the sampler should retain the full vocabulary for that row rather than attempting to scatter using an index equal to `vocab_size`.

The operation should complete normally and return sampled tokens/probabilities without corrupting the CUDA context.

### actual behavior

The sampler can generate an out-of-bounds scatter index.

For the reproduced case, the relevant logic in:

`tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py`

performs approximately:

```python
mask_to_remove = cumulative_probs >= top_p
last_index_to_keep = torch.searchsorted(
    cumulative_probs,
    top_p,
    side="left",
)
mask_to_remove.scatter_(1, last_index_to_keep, ...)
```

### additional notes

---

# 8. Additional Notes

Paste this:

````markdown
### Root cause investigation

The issue appears to be caused by the assumption that every row of the cumulative probability tensor must contain an element greater than or equal to `top_p`.

For the reproduced input, the final FP32 cumulative probability can remain slightly below `0.9999999` because of floating-point accumulation.

In the CPU isolation of the same case:

```text
top_p (float32): 0.9999998808
final cumulative probabilities:
  row 0: 0.9999999404
  row 1: 0.9999998212

searchsorted indices:
  row 0: 31999
  row 1: 32000
````

### Before submitting a new issue...

* [x] Make sure you already searched for relevant issues, and checked the [[documentation](https://nvidia.github.io/TensorRT-LLM/)](https://nvidia.github.io/TensorRT-LLM/) and [[examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples)](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.
